File: services/audit-api/src/audit/fetcher.py
"""
URL fetcher with security and rate limiting.
"""

import asyncio
from typing import Optional, Dict, Any
from urllib.parse import urlparse, urljoin
import httpx
import structlog

from audit.security import SecurityValidator
from config import settings

logger = structlog.get_logger(__name__)


class Fetcher:
    """Secure URL fetcher with rate limiting and validation."""
    
    def __init__(self):
        self.security_validator = SecurityValidator()
        self._client = None
    
    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create HTTP client."""
        if self._client is None:
            self._client = httpx.AsyncClient(
                timeout=httpx.Timeout(settings.audit_timeout),
                follow_redirects=True,
                max_redirects=settings.max_redirects,
                limits=httpx.Limits(
                    max_keepalive_connections=20,
                    max_connections=100,
                ),
                headers={
                    "User-Agent": settings.user_agent,
                    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                    "Accept-Language": "en-US,en;q=0.5",
                    "Accept-Encoding": "gzip, deflate",
                    "Connection": "keep-alive",
                    "Upgrade-Insecure-Requests": "1",
                },
            )
        return self._client
    
    async def fetch_url(self, url: str) -> Optional[Dict[str, Any]]:
        """
        Fetch URL content with security validation.
        
        Args:
            url: URL to fetch
            
        Returns:
            Dict containing content and metadata, or None if failed
        """
        try:
            # Validate URL security
            await self.security_validator.validate_url(url)
            
            client = await self._get_client()
            
            logger.info("Fetching URL", url=url)
            
            response = await client.get(url)
            response.raise_for_status()
            
            # Check content type
            content_type = response.headers.get("content-type", "").lower()
            if not self._is_allowed_content_type(content_type):
                raise ValueError(f"Content type not allowed: {content_type}")
            
            # Check content size
            content_length = response.headers.get("content-length")
            if content_length and int(content_length) > settings.max_content_size:
                raise ValueError(f"Content too large: {content_length} bytes")
            
            # Get content
            content = response.text
            
            # Check actual content size
            if len(content.encode("utf-8")) > settings.max_content_size:
                raise ValueError("Content too large after decoding")
            
            return {
                "url": url,
                "content": content,
                "status_code": response.status_code,
                "headers": dict(response.headers),
                "content_type": content_type,
                "content_length": len(content.encode("utf-8")),
                "final_url": str(response.url),
            }
            
        except httpx.HTTPStatusError as e:
            logger.warning(
                "HTTP error fetching URL",
                url=url,
                status_code=e.response.status_code,
                error=str(e)
            )
            return None
            
        except httpx.RequestError as e:
            logger.warning(
                "Request error fetching URL",
                url=url,
                error=str(e)
            )
            return None
            
        except Exception as e:
            logger.error(
                "Unexpected error fetching URL",
                url=url,
                error=str(e),
                exc_info=True
            )
            return None
    
    def _is_allowed_content_type(self, content_type: str) -> bool:
        """
        Check if content type is allowed.
        
        Args:
            content_type: Content type header value
            
        Returns:
            True if content type is allowed
        """
        allowed_types = [
            "text/html",
            "text/plain",
            "application/xhtml+xml",
            "application/xml",
            "text/xml",
        ]
        
        return any(allowed_type in content_type for allowed_type in allowed_types)
    
    async def close(self):
        """Close HTTP client."""
        if self._client:
            await self._client.aclose()
            self._client = None

File: apps/web/src/lib/keywords/fetcher.ts
import { URL } from 'url';

export interface FetchResult {
  html: string;
  status: number;
  bytes: number;
  warnings: string[];
}

// SSRF protection - check if hostname resolves to private/reserved IPs
function isPrivateIP(hostname: string): boolean {
  // Check for localhost and internal TLDs
  if (hostname === 'localhost' || 
      hostname.endsWith('.internal') ||
      hostname.endsWith('.local') ||
      hostname.endsWith('.test')) {
    return true;
  }

  // Check for IPv4 private ranges
  const ipv4Regex = /^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/;
  const match = hostname.match(ipv4Regex);
  
  if (match) {
    const [, a, b, c, d] = match.map(Number);
    
    // 10.0.0.0/8
    if (a === 10) return true;
    
    // 172.16.0.0/12
    if (a === 172 && b >= 16 && b <= 31) return true;
    
    // 192.168.0.0/16
    if (a === 192 && b === 168) return true;
    
    // 127.0.0.0/8 (loopback)
    if (a === 127) return true;
    
    // 169.254.0.0/16 (link-local)
    if (a === 169 && b === 254) return true;
  }

  return false;
}

function isValidScheme(url: URL): boolean {
  return url.protocol === 'http:' || url.protocol === 'https:';
}

export async function safeFetch(url: string): Promise<FetchResult> {
  const warnings: string[] = [];
  
  try {
    // Parse and validate URL
    const parsedUrl = new URL(url);
    
    // Check scheme
    if (!isValidScheme(parsedUrl)) {
      throw new Error('Only HTTP and HTTPS schemes are allowed');
    }
    
    // Check for private IPs
    if (isPrivateIP(parsedUrl.hostname)) {
      throw new Error('Private/reserved IP addresses are not allowed');
    }
    
    // Create AbortController for timeout
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 10000); // 10s timeout
    
    // Fetch with redirect limit and size limit
    let redirectCount = 0;
    const maxRedirects = 5;
    let currentUrl = url;
    
    while (redirectCount <= maxRedirects) {
      const response = await fetch(currentUrl, {
        method: 'GET',
        headers: {
          'User-Agent': 'Mozilla/5.0 (compatible; RIVISO-Keywords-Bot/1.0)',
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
          'Accept-Language': 'en-US,en;q=0.5',
          'Accept-Encoding': 'gzip, deflate',
          'Connection': 'keep-alive',
          'Upgrade-Insecure-Requests': '1',
        },
        signal: controller.signal,
        redirect: 'manual' // Handle redirects manually
      });
      
      clearTimeout(timeoutId);
      
      // Handle redirects
      if (response.status >= 300 && response.status < 400) {
        const location = response.headers.get('location');
        if (!location) {
          throw new Error('Redirect response missing Location header');
        }
        
        // Resolve relative URLs
        const redirectUrl = new URL(location, currentUrl);
        
        // Check redirect URL for security
        if (!isValidScheme(redirectUrl)) {
          throw new Error('Redirect to non-HTTP/HTTPS scheme not allowed');
        }
        
        if (isPrivateIP(redirectUrl.hostname)) {
          throw new Error('Redirect to private IP not allowed');
        }
        
        currentUrl = redirectUrl.toString();
        redirectCount++;
        continue;
      }
      
      // Check final status
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      
      // Check content type
      const contentType = response.headers.get('content-type') || '';
      if (!contentType.includes('text/html')) {
        warnings.push(`Unexpected content type: ${contentType}`);
      }
      
      // Check content length
      const contentLength = response.headers.get('content-length');
      if (contentLength) {
        const bytes = parseInt(contentLength, 10);
        if (bytes > 1.5 * 1024 * 1024) { // 1.5MB limit
          throw new Error(`Content too large: ${bytes} bytes`);
        }
      }
      
      // Read response with size limit
      const reader = response.body?.getReader();
      if (!reader) {
        throw new Error('No response body');
      }
      
      let html = '';
      let totalBytes = 0;
      const maxBytes = 1.5 * 1024 * 1024; // 1.5MB
      
      try {
        while (true) {
          const { done, value } = await reader.read();
          if (done) break;
          
          totalBytes += value.length;
          if (totalBytes > maxBytes) {
            throw new Error(`Content too large: ${totalBytes} bytes`);
          }
          
          html += new TextDecoder().decode(value);
        }
      } finally {
        reader.releaseLock();
      }
      
      return {
        html,
        status: response.status,
        bytes: totalBytes,
        warnings
      };
    }
    
    throw new Error(`Too many redirects (${maxRedirects})`);
    
  } catch (error) {
    if (error instanceof Error) {
      if (error.name === 'AbortError') {
        throw new Error('Request timeout (10s)');
      }
      throw error;
    }
    throw new Error('Unknown fetch error');
  }
}
